### ISSUE TITLE
Etags: solutions besides clearing/disabling cache [which does clear/block etags, sorry about the confusion]

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#179]
### Test 1

Use this [ETag Test PoC](http://lucb1e.com/rp/cookielesscookies/), type in something unique, close the browser, restart.

I do not use disk cache, but I do use memory cache (RAM). My settings on close are to clear "cache". This is the same as in the Clear recent history> dialog - where "cache" is also checked.

Note: I am not allowing JS, as it is not needed

```js
user_pref("privacy.clearOnShutdown.cache", true); // 2803 - on close
user_pref("privacy.cpd.cache", true); // 2804 - manual clearing 
```

Expected results: To not be tracked
Actual results: I failed the test. I was tracked.

### Test 2

Repeat the test but use a different string, and instead of closing the browser, simply clear your cache via History>Clear Recent History ( or Ctrl-Shift-Del )

Result: Same - I was tracked

### Test 3

~~Go to [JoDonym](http://ip-check.info/?lang=en) landing page~~

Actually this test will no longer work. JoDonym used to have an ETag test. If you cleared your cache, AFTER the landing page and THEN performed the test, the ETag would be green, otherwise it would be red - which led me to believe that clearing cache cleared the godamn cache.

## So what the heck is going on?

Clearing cache should clear the cache - right? Eg if I start a new FF session, clear the cache, load a page and look at the media elements (page info> media tab), you can see contents. In the past if you cleared cache, this would all no longer be shown.

I **KNOW** memory cache used to get cleared from my experience with JoDonym and from looking at items in the cache - both with FF's UI, and with extensions in the past. I have only ever used memoery cache.

~~So either some pref is causing a block on this (eg FPI) or Mozilla decided that memory cache no longer needed cleaning?~~

Class, discuss!

**Comment by Thorin-Oakenpants** [comment]
Here's an old JoDonym to show what I'm talking about. It seems `Authentication`, and `Cache (E-Tags)` are no longer shown/run

![oldjodonym](https://user-images.githubusercontent.com/16656956/28538260-c1385a50-7101-11e7-9194-2a3a99243286.png)

**Comment by Atavic** [comment]
[Cookieless Cookie Fake Test](https://anonymous-proxy-servers.net/blog/index.php?/archives/400-Cookieless-Cookie-Fake-Test.html): The test uses the IP address for session tracking and the user agent but not ETags.[

This test](https://ochronus.com/tracking-without-cookies/) should use ETags, instead.

**Comment by Just-me-ghacks** [comment]
http://lucb1e.com/rp/cookielesscookies/

Results:
browser.cache.memory.enable TRUE & same IP - I was tracked.
browser.cache.memory.enable FALSE & same IP - I was not tracked.
browser.cache.memory.enable TRUE & different IP - I was not tracked.

This "test" could be fake! We need a more reliable ETag test.

**Comment by Thorin-Oakenpants** [comment]
As mentioned in the FPI sticky - https://bugzilla.mozilla.org/show_bug.cgi?id=1371651 - FPI & about:cache (for a fleeting moment before I looked it up, I thought this might be a smoking gun - but it's just the about:cache page, not the actual cache). However, these tests should be done in a nilla profile

**Comment by Thorin-Oakenpants** [comment]
PS: the code is here on github - https://github.com/lucb1e/cookielesscookies

**Comment by Thorin-Oakenpants** [comment]
I want a medal ... 14 years ... https://bugzilla.mozilla.org/show_bug.cgi?id=231852

**Comment by Thorin-Oakenpants** [comment]
bit of a recent'ish discussion going on here and you have to read thru the chaff to get to the wheat - https://groups.google.com/forum/#!topic/mozilla.support.firefox/ZE0_3tKGlUU (yeah yeah shitty google docs - go use a secondary browser like I do)

About 53.7% of way down
> The purge-on-exit settings that Firefox affords in its GUI are not deleting its web cache on exit.

**Comment by Thorin-Oakenpants** [comment]
OK, I thought about this and now someone mentioned (see last post for link) - modify headers. Well, I listed a WE one a few weeks ago .. its still installed ...

> I blocked Etags by using the extension Modify Headers and added a filter 
for ETag.  Now my browser does not keep or return ETags any more. 

But he doesn't say any more than that

**Comment by Atavic** [comment]
Noscript. NoScript's ABE ([Application Boundaries Enforcer](https://noscript.net/abe/)) allows stripping etags.

[Secret Agent](https://www.dephormation.org.uk/index.php?page=81) generates spoof ETags headers with random values, preventing ETags being misused for tracking.

Run a local proxy as [Privoxy](http://www.privoxy.org/).
Privoxy can [block ETag headers](https://www.privoxy.org/user-manual/actions-file.html#CRUNCH-IF-NONE-MATCH).

**Comment by 2glops** [comment]
Tor Browser 7.0.2 (vanilla)
TB with stored text > tracked by number of visits but not by stored text
TB with stored text + new identity > not tracked

**Comment by Thorin-Oakenpants** [comment]
Secret Agent isn't even on AMO. I've already scoped it out. And I think `noise` is worse than `block`. Its a stupid solution IMO to generate these headers for everything. I know about Privoxy, I used to use it (Windows). In fact, I've been meaning to add it to my router (DD-WRT'd and all). I want solutions for Firefox, in Firefox.

This [Modify Header Value (HTTP Headers)](https://addons.mozilla.org/en-US/firefox/addon/modify-header-value/?src=search) I have tried and tried and tried and cannot get it to work (and yet no one seems to think its broken). Github page listed in addons sticky. I cannot get it to block `ETag` as far as I can tell. This looks like a very good extension, but I'm f**ked if I can make it work properly.

Now NS's ABE - well? Please provide. NoScript>Options>Advanced>ABE>User ... add this rule - and provide the rule ( https://noscript.net/abe/abe_rules.pdf - this is over my head, maybe someone can craft a rule for blocking/stripping outbound ETag HTTP Headers )

**Comment by Thorin-Oakenpants** [comment]
Edit: an old read but might explain something - transparent proxies - http://lifecs.likai.org/2011/07/kissmetrics-and-life-of-etag.html

**Comment by Atavic** [comment]
Installed Modify Header Value (HTTP Headers) on FF ESR 52.0

In HTTP Headers options I created a new entry with:
URL `*`
Header Name `If-None-Match:`
Header Value `*` (maybe useless)

Then I picked remove.

By testing the link [here](https://github.com/ghacksuserjs/ghacks-user.js/issues/179#issuecomment-317541105) with HTTP Headers Off:

The text I entered remains in a new session (after closing and reopening FF)

With HTTP Headers On the text is lost, but the number of visits still remain.

**Comment by Forsaked** [comment]
@Atavic This way every other Website breaks its layout/design.

**Comment by Atavic** [comment]
You can whitelist sites, deactivate your filtering choice or launch a second browser with this WebExtension on [AMO](https://addons.mozilla.org/en-US/firefox/addon/external-application/), part of these [projects](https://add0n.com/open-in.html) by [Andy Portmen](https://github.com/andy-portmen/native-client).

The HTTP Entity Tag is used to reduce the server bandwidth when a resource was previously stored in the cache. [Abusing the ETag](http://en.wikipedia.org/wiki/HTTP_ETag#Tracking_using_ETags) values allows tracking of the browser: the ETag value is unique as an MD5 or SHA1 hash. I can either strip the hash value from every 'If-None-Match' header value with Privoxy or spoof it with [Secret Agent Add-on](https://github.com/ghacksuserjs/ghacks-user.js/issues/179#issuecomment-317558779): its spoofing feature adds a random hash to outgoing requests, while its whitelist feature allows you to specify the trusted hosts. For Privoxy issues there's an active [forum](http://www.prxbx.com/forums/forumdisplay.php?fid=49).

**Comment by zymase** [comment]
I haven't gone through the tests mentioned above but maybe is it worth remembering the LocalCache feature which isn't cleared by Firefox's cache delete function.

LocalCache is all in the user's profile webappsstore.sqlite file.
LocalStorage is handled in the user's about:config with:
// user_pref("dom.storage.enabled", true); // default=true

Closing Firefox -- deleting webappsstore.sqlite -- restarting Firefox brings a fresh, new, immaculate (pure, virgin etc) webappsstore.sqlite (brand new is it). But the user will have lost all data kept in LocalStorage. I've never experienced any issue with that.

**Comment by Thorin-Oakenpants** [comment]
Hmmm .. where did zymase's comment go? .. Yup, this is testing ETags zymase, there is no JS, no cookies, no dom etc etc

**Comment by zymase** [comment]
LOL!

> Hmmm .. where did zymase's comment go? .. Yup, this is testing ETags zymase, there is no JS, no cookies, no dom etc etc

That's why I had removed my comment. Thanks for not including in your brief recall of my shame the fact I love chocolate and mini-skirts!

Yeah, I had commented too quickly, suggesting the LocalStorage to be an explanation of the cache issue detailed by this thread. I then took the time to read extensively, and at this time I've installed The _Secret Agent_ add-on (from developer's page mentioned above by Atavic, when unavailable on AMO).

Hey! I'm a secret agent man as sung by Austin Powers; yeah! -- Seriously, the add-on is most interesting, testing it right now as I said. One thing is sure, the cache issue revealed, exposed, applied with the  _**ETag Test PoC**_ is obsolete with **_Secret Agent_**.

**Comment by Thorin-Oakenpants** [comment]
There's something whacky about how we're doing the tests or that test page. I found i am not tracked when I run CCleaner, so to me that means the test is using some form of persistent disk data

**Comment by zymase** [comment]
After some time testing the _Secret Agent_ add-on I'm finally keeping it but will be using only its **_eTags & Caching_** feature because it seems to handle the eTag intrusion correctly by spoofing it.

_Secret Agent_ has several other features but too many sites get problematic when these features are enabled. Of course the add-on has a whitelist which is flawless but I've had to whitelist domains, sites only because of the features I then removed. But whitelisting a site will then stop filtering its eTags as well. 

In other words, only eTag spoofing and no sites (up to now) need to be whitelisted 

_Secret Agent_ then but only for eTag spoofing, which it does well, and which resolves the very concern of this thread, as I see it.

One last thing: Secret Agent is displayed with a toolbar button on its dedicated toolbar, positioned right under the urlbar and reserving the screen width for only its toolbar button followed by description of the add-on's status (what it's doing, is it on or off). Removing the toolbar button is possible, can then be positioned anywhere and the toolbar itself can be hidden with right-click on the tabs toolbar then uncheck _Secret Agent toolbar_. Of course Secret Agent toolbar button alone will not be followed by description of status as when it is on its toolbar, but hovering it provides the information and right-clicking it opens a drop-down menu for whitelisting the site, disabling/enabling Stealth Mode and access to Options.

Done this way, seems interesting to me.

**Comment by passboltUser** [comment]
I don't like the idea of needing another addon (and why isn't it in AMO anyway? Not that I would use it if it were). 

As I and the author understand it, cache in general is the problem: "[...] _cache tracking issues_ [...]" and "[...] _I currently have no straightforward answer since cache tracking is virtually undetectable,_ [...]" .
I don't understand, why is cache not cleared if Firefox is closed?
This tracking seems to be disabled if `browser.cache.memory.enable` is set to `false` (https://github.com/ghacksuserjs/ghacks-user.js/issues/179#issuecomment-317534877).
I would prefer deleting the cache if Firefox is closed. This seems not to work? (https://github.com/ghacksuserjs/ghacks-user.js/issues/179#issuecomment-317547307) What's the status on that?

**Comment by Thorin-Oakenpants** [comment]
> What's the status on that?

I have no idea. I want to get the FF55 changes finished. When I get back to this (do not know when), because it is intriguing, I will do more thorough analysis and testing. Better/more ETAG test sites, and also comparing the results using various FF releases. Whatever it is, my gut instinct is that since CCleaner cleans it, it must be local - and since all previous advice in forums, etc has been to clear "cache" - this no longer seems to work (on that one site) - hence I will setup and test old FF portables

**Comment by passboltUser** [comment]
@Thorin-Oakenpants Thanks for your work then. This is indeed intriguing. I think this must be local indeed, because if I use a second profile, the tracking-cookie isn't there. I wouldn't bother with ETags, because the cache is the generic issue here and who knows that other techniques are used using cache. The real issue is, as you pointed out in your initial post, is that the cache is not being cleared -- if it's set to be cleared -- when FF is closed. I'd personally would want the cache to be cleaned when FF is closed (expected behavior), or I would disable cache altogether (not the preferred solution).

**Comment by zymase** [comment]
Another fingerprinting test, and this one is the toughest I know : [Cross-browser fingerprinting test 2.0](https://fingerprint.pet-portal.eu/)

**Comment by 2glops** [comment]
@zymase 
The FP test don't start, blocked by uBO (dynamic filtering):
https://fingerprint.pet-portal.eu/javascript/evercookie/evercookie.js

**Comment by zymase** [comment]
Of course it's always possible to add a uBlockO filter to block Cross-browser fingerprinting test 2.0, such as `/evercookie.js$important` but that solves a test, an effect and not the cause. Moreover I doubt a site using evercookies would name their script [evercookie.js](https://fingerprint.pet-portal.eu/javascript/evercookie/evercookie.js).

**Comment by 2glops** [comment]
@zymase 
You're right and I just deleted my comment after a short time while you were answering. 
Gulty ! That will not happen again.

**Comment by zymase** [comment]
@2glops it happens to everyone! If deleting a comment is authorized it's for a good reason.

So, we can use medicine to block evercookie.js (namely) but we cannot cure the fingerprinting disease :)

**Comment by Thorin-Oakenpants** [comment]
Bandaids are stoopid (let the bastards bleed), playing whack-a-mole is re-DICK-ulous and a waste of resources, the cure for cancer is hard, space is mindbogglingly huge and vast (do not eat the fairy cake!), and tests should be under worse case scenarios (I can disable JS for everything forever, therefore I win this contest - and to beat all server side FP, I will never connect to any servers, YAY, I win again)

**Comment by Thorin-Oakenpants** [comment]
some links for reference
* the meantime exploit - http://sourcefrog.net/projects/meantime/ - scroll down
   * ETag and  Last-Modified headers
* [231852](https://bugzilla.mozilla.org/show_bug.cgi?id=231852) - only 14 years old

**Comment by crssi** [comment]
About ETags and performace...
To get rid of ETags, I have set:
`user_pref("browser.cache.memory.enable", false);`
`user_pref("browser.cache.memory.capacity", 0);`

No ETags, but I don't see any performance issues too.

**Comment by crssi** [comment]
^^Is any of those master switch and I don't need the other one?

**Comment by Thorin-Oakenpants** [comment]
Yes, we already know that disabling memory cache also stops etags, because there is NO cache of any sort, so if-match-none headers are never sent (or so I assume). But browsing around with zero memory cache (we already have no disk cache) is a bit extreme. I am looking for a better solution.

What I am more concerned about right now is that clearing everything (manually or on close) does not clear the etags - ****not even in private browsing mode**** (are you hearing me Mozilla people who work on PB mode stuff) - so that is problem no. 1

Problem no 2 is that I want to remove outgoing (I think it's outgoing) If-None-Match headers and I can't - I am not a header specialist like earthlng, so I probably f**d it up. I want to use my cache if possible and not force everything to re-download.

As for your question, I would say that the first pref is a master switch, but don't quote me on it

**Comment by crssi** [comment]
As I say, I don't see any performance issues. Why don't you try it?

**Comment by Thorin-Oakenpants** [comment]
> Why don't you try it?

Because I am looking for a better, more robust solution for all users

**Comment by crssi** [comment]
That is OK and understandable. But I have asked why don't YOU try it. I haven't implied anything further, at least not to include it into public version.

**Comment by Thorin-Oakenpants** [comment]
I'm running with no memory cache now. Sometimes I think I feel a difference (longer loading times), but since I never load heavy sites or any media (vids, audio), and never really have more than 6-10 tabs open at a time, I do not think its a problem. It's also not so bad considering all the XSS & JS shit that is blocked

What I would like is some extension that indicated when etags are present  - would need some sort of filtering perhaps to remove false positives

As for [modify header value http headers](https://addons.mozilla.org/en-US/firefox/addon/modify-header-value/) - this extension sucks. Comments seem to confirm same as me that 

> The domain checkbox has no affect. Still affects the whole domain instead of just URL homepage even when unchecked. The modify and remove checkboxes also have no affect. The header don't get modified and remove doesn't work. The domain, modify and remove options need fixing.

Earthlng & co: where is a decent header extension that can modify on the fly as per some simple rules eg

![etags-be-gone](https://user-images.githubusercontent.com/16656956/33699535-9ba4a9b0-db78-11e7-9fcb-cb294f5539d3.png)

**Comment by earthlng** [comment]
you can try this one: https://addons.mozilla.org/en-US/firefox/addon/header-editor/
I have my own home-made WE to do these kinds of things but I'm not gonna release it anytime soon. (or maybe ever)
You can't remove If-None-Match though because it's one of the headers that gets added *after* WEs had their hands on them. But removing the ETag **Response** Header should work.

**Comment by Thorin-Oakenpants** [comment]
^^ its in tolkeinese mate .. does it come in English ?

**Comment by Thorin-Oakenpants** [comment]
@earthlng OK, installed, added cookie exception and it has set up it's IDB. Ready to test it with memory cache back on. I see it has an import function. Write me a rule buddy :)

**Comment by crssi** [comment]
For me too...pleaseeee :)

**Comment by earthlng** [comment]
The code doesn't use .toLowerCase to compare headers but I think existing headers are now always lowercase, meaning you need use lowercase when you want to remove a certain header. BUT if you want to **add** one you should probably use the correct casing, fe. if you always want to add a Referer Header, name it 'Referer' not 'referer' or you might make yourself pretty unique because nobody else sends a lowercase referer header!
Anyway, I think this should do:

![etag](https://user-images.githubusercontent.com/25846669/33771513-86be145a-dc31-11e7-90bd-c186b9b9df05.png)

**Comment by Thorin-Oakenpants** [comment]
cheers, will do once I sort out this 0850a+b mess

**Comment by crssi** [comment]
Thank you @earthlng, this seems to get rid of etags with memory cache enabled.

Nice

**Comment by Thorin-Oakenpants** [comment]
well that works AFAICT .. https://httpbin.org/ <-- we need to add this to the wiki

**Comment by crssi** [comment]
Test at [http://lucb1e.com/rp/cookielesscookies/ ](http://lucb1e.com/rp/cookielesscookies/ ) also passes in desired way.

**Comment by Thorin-Oakenpants** [comment]
thanks earthlng - I had a feeling it could be done because of what that guy wrote ages ago about header fiddling. Which is why I persisted in leaving this open and finding a solution. Shame I used a shit extension to test it.

crssi .. aren't you glad I left this open for 4 1/2 months

**Comment by earthlng** [comment]
If-None-Match sends the Etag if one was received but since you're now removing it there's nothing to send. And you can still profit from caching because of the Last-Modified header. That *could* also theoretically be used to fingerprint by sending unique dates for each user and request but I don't think anyone does that (yet?!). If you want to remove that too you might as well just disable caching again.

**Comment by earthlng** [comment]
> Shame I used a shit extension to test it.

comments like that are part of why I'll probably never release an extension again. Just because people don't understand how it works and what it does, doesn't make it shit. Just sayin'

**Comment by Thorin-Oakenpants** [comment]
^^ but it didn't work with editing/removing/whatever in all sorts of configs for etags (I only tried the if-none-match a few days ago when I had time to revisit this). I don't profess to being an expert, but I'm not an idiot. It did not work, and others said the same thing (although we could all be wrong and are dummies)

**Comment by Thorin-Oakenpants** [comment]
> because of the Last-Modified header. That *could* also theoretically be used to fingerprint by sending unique dates for each user and request but I don't think anyone does that (yet?!)

Yeah, that's what I meant - it's specifically mentioned as a threat in the http://sourcefrog.net/projects/meantime/ link I gave. As you said, disable memory cache is the only complete solution for now - except for the legacy extensions `secret agent` (not on AMO) and `RAS` (I think it has something about etags), but not sure if they handle this Last-Modified header

**Comment by earthlng** [comment]
It didn't work because FF adds that header after changes made by WEs. Not knowing that doesn't make you an idiot but ranting about how shit it is makes you kind of an asshole. It even clearly states

> Modify Header Value can add, modify or remove an HTTP-**request**-header for all requests on a desired website or url.

obviously if you try to modify a **response** header it won't work. There's a big difference between
"it doesn't work - this is shit" and "I can't get this or that to work - can you help, please", wouldn't you agree?

**Comment by Thorin-Oakenpants** [comment]
But but butttt ... this one works, so I fail to see the diff

**Comment by earthlng** [comment]
Re: last-modified - you could use a custom function to add or remove a random amount of seconds or minutes but since I doubt that this FP technique is commonly used, it would make you stand out more on most websites than just sending back the original value.

**Comment by Thorin-Oakenpants** [comment]
I am pushing Arthur to take up bugzilla [231852](https://bugzilla.mozilla.org/show_bug.cgi?id=231852) where I think they could do some real damage control

**Comment by crssi** [comment]
Sure I am. :)

**Comment by Atavic** [comment]
>Why aren't the Dephormation and Secret Agent add ons published on Addons.mozilla.org?

I don't subcribe to the view that there can only be one marketplace for computer software.

**Comment by Atavic** [comment]
>Why aren't the Dephormation and Secret Agent add ons published on Addons.mozilla.org?
>I don't subcribe to the view that there can only be one marketplace for computer software.

^From dephormation.org.uk FAQ

RAS has a `Send Spoofed If-None-Match headers` [option](https://github.com/dillbyrne/random-agent-spoofer/wiki/Headers-Tab-Usage#standard-options)

**Comment by earthlng** [comment]
https://github.com/dillbyrne/random-agent-spoofer/wiki/Headers-Tab-Usage#standard-options

> I would like to improve on this so we eventually send nothing but have not been able to get a better solution thus far.

Well, I just gave you that solution. You can of course "spoof" the If-None-Match header indirectly by changing the incoming Etag header but AFAIK you can't tamper with the outgoing If-None-Match header directly. This is at least the case for Webextensions, it might have been different with legacy addons.
And ...
> Random Agent Spoofer
Note: This addon is no longer being maintained as it uses the legacy addon sdk. I will not be porting it to web extentions due to certain incompatibilities with the web extensions API. It will stop working in firefox 57

**Comment by Thorin-Oakenpants** [comment]
https://github.com/ghacksuserjs/ghacks-user.js/wiki/4.2.4-Header-Editor

**Comment by Thorin-Oakenpants** [comment]
https://anonymous-proxy-servers.net/blog/index.php?/archives/400-Cookieless-Cookie-Fake-Test.html

So the PoC we were using seems to be a bit dodgy, as we suspected. Haven't read the link fully, and it does not invalidate the ETag tracking concept or the current header editor solution

**Comment by claustromaniac** [comment]
@Thorin-Oakenpants I read that entry some time ago but after skimming the source code I couldn't find anything suspicious. In fact, the source was surprisingly simpler than I expected. Also, some of the buggy behaviour is explained in the very test page, below.

I wouldn't go as far as to call it a fake test, I'd just say it's kinda buggy, unreliable, probably in part because of it's simplicity, and the fact that it seems like it wasn't updated in four years.

Then again, I am by no means an expert on the subject.

**Comment by Thorin-Oakenpants** [comment]
I can't explain the IP thing, maybe that's caused internally by FF. But for sure with the only change being block the etag response header, it works. The IP thing is kinda moot - nice if you can flick between ips every 10 minutes like JoDonym, but not practical for most people

> Then again, I am by no means an expert on the subject.

Me too. Earthlng berated me and I had to resort to a night of beersies for cold comfort

**Comment by claustromaniac** [comment]
I just skimmed the source for a second time and it seems the IP and user agent are used as seeds to first generate a unique ETag for each user, but using a pseudo-random number instead should produce mostly the same results in that context. It shouldn't be relevant to the tracking mechanism.

Something else must be in play. It would be nice to find out what that is, exactly. Not just out of curiosity, it could be useful. For starters we could test from different browsers, to see if it is a Firefox-only thing or not.

Of course, summoning some gurus here would hopefully save us the hassle of testing............. *starts drawing the summoning circle just in case*

**Comment by Thorin-Oakenpants** [comment]
> starts drawing the summoning circle just in case

wait .. is that meant to be salt? and do I stay on the inside or outside of the circle?

**Comment by Gitoffthelawn** [comment]
Here are my latest test results using the ***cookieless cookies*** test page at http://lucb1e.com/rp/cookielesscookies/  (note: read the whole page... there is good information within).

On Firefox v58.0beta10 (tweaked), it always shows 2 visits, and does not retain the text.  Refreshing the page does not increment the counter, and the store string is not maintained.  I'm not sure why it always shows 2 visits, but the counter not incrementing, and the string not being stored, are both encouraging.  Anyone have similar results and ideas on how to interpret them?

On Firefox Klar/Focus (Android) v2.5, the counter increments.  Clearing the cache one time has no effect on the next time the page is loaded.  However, loading the page one more time (or clearing the cache again) reduces the counter to 2 and removes the stored string.  Again, I'm interested in your feedback on interpreting these results.

**Comment by earthlng** [comment]
What seems to be the problem here guys? That it always says "Number of visits: 2" ?
It's because his sessions folder is stacked full with previous sessions because it never cleans up.

**Comment by earthlng** [comment]
> Clearing the cache one time has no effect on the next time the page is loaded.

that's because it takes a short while to clear the cache

**Comment by claustromaniac** [comment]
@Thorin-Oakenpants 

> do I stay on the inside or outside of the circle?

On the outside, unless you don't mind lending some blood for the ritual.

**Comment by claustromaniac** [comment]
Wow I see a lot of confusion going on [here](https://www.ghacks.net/2017/12/09/a-solution-to-etag-tracking-in-firefox/) lol. I tried to explain what I found in the code and why it's theoretically acceptable to make ETags out of remote IP + user agent, but the explanation kinda got buried down there, among the many test results and different claims.

I should've been clearer from the start. Regardless, I still can't explain what's going on with the test once the client changes IP.

**Comment by Atavic** [comment]
Related: https://github.com/lucb1e/cookielesscookies/issues/18

**Comment by Thorin-Oakenpants** [comment]
@carmenbianca  - I have three points
- we/I were wrong about cache not clearing etags (kane gives instructions to show it being cleared in the dev console/inspector or whatever)
- the tracking by etags is fully documented and while the PoC is flawed, the threat remains
- since clearing cache does work, users have three options
   * Option1: don't use disk or memory cache - set & forget
   * Option2: clear cache periodically - you can do this automagically eg with uM - but still leaves repeat visits open during that window of time - image people clearing cache only on close and their browser isopen for hours even days at a time. Doing it manually is not suitable either. Also, if you're clearing cache every 10 minutes, why bother, just disable cache then
   * Option3: block the etag header, allows using cache - set & forget - makes you look like the same as option1

As for the point about FPing and what the real world threat is, is debatable. This is a FP data point that  you can't just block like JS, but at least it is only two values for those with no cache or blocking etags - on or off. We're always off. Other people with etags could potentially be unique. I would rather be in the 2-3% than risk being uniquely ID'ed

--
PS: I read Kane's comments in the first comment section a number of times. Like you said, even if you changed IP, the code should still read your previous etag value and keep incrementing. At the end of the day the PoC used without IP changes, still shows, I think, a valid result. Regardless - the concept is KNOWN to work, so I'm not worried about trying to prove it anymore.

PPS: I removed the PoC from the list of test sites. If we add it back in we would need to clarify the steps to test

**Comment by Thorin-Oakenpants** [comment]
> and that this PoC does not rely on the client's IP staying the same by design.

That's almost my take as well - do the test without changing IP and the PoC holds (kinda because the same etag is used which confuses things). It should be a unique etag for every visit if they do not already have an etag

- 1st Visit: you have none, so etag assigned `123`
- 2nd visit: etag read back `123`
- clear cache
- 3rd visit: you have none, so new etag assigned `456` <-- instead you are reassigned the same etag `123`

I'm thinking this thru, and on a new PoC might have to allow JS just for the display side of things
- 1st visit: you  have none, so etag assigned from random guid generator which I will call `A`
   * page displays "you have no etag from us, now you do, it is `A`
- close browser (no clearing of cache), or close tab, or refresh
   * you've been here before, you unique etag is `A`
- change ip (do not clear cache) and revisit
   * you've been here before, you unique etag is `A`
- clear cache and revisit
   * page displays "you have no etag from us, now you do, it is `B`

Obviously everything else (cookies, localstorage/idb, referers, etc) should be eliminated fro the equation. I'm not a wev dev, but surely this is not hard to put togther

**Comment by crssi** [comment]
@Thorin-Oakenpants It this typo?

> Option1: don't use disk **or** memory cache

and should be:

> Option1: don't use disk **and** memory cache

**Comment by Thorin-Oakenpants** [comment]
yeah .. don't use either

**Comment by carmenbianca** [comment]
@Thorin-Oakenpants Hi!  I don't think you meant to @ me? I'm not really sure what you're talking about :)

**Comment by Thorin-Oakenpants** [comment]
sorry @carmenbianca ... slip of the arrow keys .. I meant @claustromaniac just below you .. sorry :)

**Comment by vertigo220** [comment]
Ok, so after spending quite a bit of time testing Header Editor, Secret Agent, and, to a lesser extent (since it only deals with 3rd-party ETags, so it doesn't seem to have any effect here), Privacy Possum, I'm at a loss. I've read this whole thread and about half the comments so far at [https://www.ghacks.net/2017/12/09/a-solution-to-etag-tracking-in-firefox/](url), and I'm not sure how to test this. I set up a rule in Header Editor and tested it on [https://lucb1e.com/rp/cookielesscookies/](url) (PoC?), and the only way I've been able to get that site to "forget" me is to either clear cache (thanks to @earthlng for their comment about it taking a while, as I was also struggling to figure out why sometimes this worked and sometimes it didn't) or with Secret Agent*.

After reading everything so far, it seems the lucb1e site doesn't work properly, and it sounded like Pants was going to remove it as a test site, but it's still listed in section 4.2.4 of the wiki. As for the [other site](https://httpbin.org/) listed there, I have no idea how to use that. I also tried the [site](https://ochronus.com/tracking-without-cookies/) @Atavic posted above, but it doesn't seem to do anything, and the iframe doesn't seem to load properly:

![image](https://user-images.githubusercontent.com/12024557/40280576-a74bda96-5c23-11e8-9f6a-e78248e4ca1b.png)

@Thorin-Oakenpants - I pointed out to you elsewhere that you had misspelled JonDonym on the Appendix A of the wiki, but I see you've spelled it the same (JoDonym) everywhere else as well, including the ghacks article comments. Not important enough to go tracking them all down to fix them, but just wanted to mention it again in case you didn't realize how widespread it was. 😉 And of course, thanks for yet again digging into this stuff.

*As far as Secret Agent is concerned, I have no issue with it not being on AMO, especially considering how Mozilla handles (dictates) certain things with it, such as what's allowed now regarding XUL addons. That said, while the addon certainly seems to work well for this, and has lots of features, there's a _lot_ of room for improvement with it that I see. Unfortunately, it doesn't seem to be under active development (and I seriously doubt that will change with the death of XUL addons) and the dev doesn't seem open to input, so I'd just as soon figure out an alternative method for dealing with ETags.

**Comment by Thorin-Oakenpants** [comment]
> your links were fubared

what you had
`[https://www.ghacks.net/2017/12/09/a-solution-to-etag-tracking-in-firefox/](url)`
what I changed it to
`[DISPLAY TEXT](https://www.ghacks.net/2017/12/09/a-solution-to-etag-tracking-in-firefox/)`

> PoC?

Proof of Concept

> https://lucb1e.com/rp/cookielesscookies/

The test site is slightly "flawed" in that it uses the IP address to spawn a unique id. This caused confusion. If you do the test without changing IP, the PoC holds true.

> alternative method for dealing with ETags

There is no alternative. Either don't use any cache (disk or memory) or clear it constantly (which still leaves x minutes of overlap). Or block the header. End of.

To properly determine if it's working, all you need to do is use the network dev tools in FF - see https://www.ghacks.net/2017/12/09/a-solution-to-etag-tracking-in-firefox/#comment-4306294 from Kane

**Comment by vertigo220** [comment]
> There is no alternative. Either don't use any cache (disk or memory) or clear it constantly (which still leaves x minutes of overlap). **_Or block the header._**

This is what I meant by an alternative, i.e. an alternative to Secret Agent. The problem is, I can't figure out if Header Editor is working the way it's supposed to, since the PoC uses the IP to seed the tag, and therefore isn't really testing ETags specifically. I just followed Kane's instructions and either it's not working or I'm not understanding it. I loaded the page, checked the header info, saw no If-None-Match line in the request headers, reloaded the page, the line was there, cleared the cache, waited, reloaded the page, the line was gone (due to clearing cache), reloaded again and it was back.

**Comment by claustromaniac** [comment]
> and therefore isn't really testing ETags specifically

If I interpreted the source correctly, it is. The fact that the PoC doesn't hold when you change your IP or user-agent is a side effect of how the former generates the unique ETag to track you. **As long as you don't change your IP or user-agent**, you can still use it for the purpose of checking if your browser is sending If-None-Match headers or not.

I would suggest you to make sure nothing in your setup is changing your user-agent. Also, make sure to reload the page to see updated results. From the PoC:

> To demonstrate how this works without having to use Javascript, I had to find a piece of information that's relatively unique to you besides this ETag. The image is loaded after the page is loaded, but only the image contains the ETag. How can I display up to date info on the page? Turns out I can't really do that without dynamically updating the page, which requires javascript, which I wanted to avoid to show that it can be done without.
> 
> This chicken and egg problem introduces a few bugs:
> - All information you see was from your previous pageload. Press F5 to see updated data.
> - When you visit a page where you don't have an ETag (like incognito mode), your session will be emptied. Again, this is only visible when you reload the page.
> 
> I did not see a simple solution to these issues. Sure some things can be done, but nothing that other websites would use, and I wanted to keep the code as simple and as close to reality as possible.
> 
> Note that these bugs normally don't exist when you really want to track someone because then you don't intend to show users that they are being tracked.

**Comment by vertigo220** [comment]
Not changing my IP or user agent, and I'm refreshing the page over and over trying to figure this out. And I guess I'm more confused than I thought, since I was understanding the issue to be that, since it uses your IP to generate the ETag, if the IP _doesn't_ change it causes the same ETag to be generated again, making it appear as though it was never blocked, when in fact it's just replaced by an identical one.

**Comment by Thorin-Oakenpants** [comment]
Clear everything (or just cache). Load the PoC
 * NOTE: you get assigned an ETag in the `response` header
 * NOTE: there is no `If-None-Match` `request` header
![pic1](https://user-images.githubusercontent.com/16656956/40281454-785696c6-5c51-11e8-94b1-985f3dc8755d.png)

Reload the page (do not clear anything)
 * NOTE: the ETag in the `response` header is the same (for this test we need to make sure it is the same due to the way it is being generated: i.e IP address and we think UA - this is a red herring, the PoC works if we eliminate these from the equation, as claustro said)
 * NOTE: it is now being sent back the ETag to track you
![pic2](https://user-images.githubusercontent.com/16656956/40281462-8ebd3e38-5c51-11e8-893a-2af4c43b2ffb.png)

What this test above shows is that ETags can be used to track you when they send back a `If-None-Match` `response` header

---

I am using the ETag rule for Header Editor as per the wiki.
![pic3](https://user-images.githubusercontent.com/16656956/40281533-bc638940-5c52-11e8-8d75-582dc4e760b6.png)

Now what I want to know, is in my test above, does the network developer tool in the second pic above show what FF actually SENDS, because if so, then the rule does not work :confused-as-f:

TL;DR: the PoC works if we eliminate a couple of variables, the chosen solution is correct (do not send `If-Non-Match`). The question is: is the header editor rule actually working?

**Comment by Thorin-Oakenpants** [comment]
man I'm confused. We're modifying the response etag header, not the request if-none-match header? right?

**Comment by Thorin-Oakenpants** [comment]
re: JonDonym - searched repo for "jodonym" - found one instance (excluding issues) - fixed it :+1:

**Comment by Thorin-Oakenpants** [comment]
Load https://httpbin.org/#/Response_inspection/get_etag__etag_
- scroll down to `Response inspection`
- click `[GET] / etag/{etag}`
- click `Try it out`
- something something I have no idea how to use this now
- click the blue `Execute` bar
- view results

Reopening this until someone shows me how to test that the etag header is blocked

**Comment by Thorin-Oakenpants** [comment]
> Enable the rule in Header Editor, clear the cache for the site (with shift+click on the reload button) and re-test. This time you'll see that the second time you reload you will still get a 304 status code for the image (it will be cached) but you won't have any if-none-match in the request headers.

Nope. I am still getting the if-none-match request header

**Comment by claustromaniac** [comment]
Not sure what to tell you, it sure works for me. You might need to reload a couple of times to get the 304 status code. Other than that, maybe use the proper casing in the Header Editor rule for the etag header? As in ETag instead of etag? I think it shouldn't matter but.. who knows. 

Also keep in mind that whatever image you use to test needs to have a decent time to live, otherwise the server will refresh the image in your cache more often and that will mess with this test. 

Javascript could mess with that too, make sure it's disabled for that site for the purpose of testing.

**Comment by Thorin-Oakenpants** [comment]
Its definitely got the 304 - I was using the `remove-etags.png`
![step2](https://user-images.githubusercontent.com/16656956/40282497-d2d7696e-5c5f-11e8-9d01-6735eae88fbc.png)

Earthlng already mentioned the case, and I have an open ticket at header editor which seems to have been forgotten

> it sure works for me

All 17 items marked as 304 have the if-none-match header

If it works for you, what exactly is the case in your rule?

**Comment by Thorin-Oakenpants** [comment]
changing case to `ETag` solved it. BUT .. @earthlng said this is a bad idea (source: somewhere) since `etag` is more common/normal?????

**Comment by Thorin-Oakenpants** [comment]
^^ `https://github.com/FirefoxBar/HeaderEditor/issues/45`

**Comment by claustromaniac** [comment]
Weird. Both `Etag` and `etag` seem to work on my end. Sounds like it is unrelated. Maybe it's just one of those weird timing things that probably happen on the server side. Who knows.

Wait a while and try again with `etag` if you're curious. I don't know lol

**Comment by Thorin-Oakenpants** [comment]
are you using the user.js basically as is? wondering if it's a pref thing, like SSL session ticketing = false, or something

**Comment by claustromaniac** [comment]
> are you using the user.js basically as is?

I'm using mostly minimal overrides on this profile. I enabled most of the Personal section, enabled disk cache, and relaxed the shoulder-surfing and related stuff I don't need to worry about.

**Comment by vertigo220** [comment]
@claustromaniac's method worked for me. Glad that's finally sorted out, been working on it on and off for a few days now. @Thorin-Oakenpants, I had the opposite problem as you at first: the tag wasn't present even after disabling Header Editor and reloading the page again. I wonder if your problem is related but in reverse. Maybe try disabling the rule, shift+F5, F5, enable the rule, shift+F5, F5. That's basically how I got it to _show_ the tag then discard it again, and it seems to be working now. I guess there's some interference somewhere in there with all the back and forth and reloading and cache clearing.

**Comment by Thorin-Oakenpants** [comment]
it might work for you on *that* site (or some sites) but will not work on *all* sites - the problem is the matching case - so this problem is not resolved IMO

meanwhile, I have two rules now, but again, its not a solution (and I do not know if this works properly), and I want to test a lot more in a nilla profile etc.

I need test sites that work with `ETag` only and one that works with `etag` only etc. But they still doesn't explain claustro's statement that both worked for him. I wish this was just built into RFP to be honest

**Comment by claustromaniac** [comment]
I'm suspecting it might be an issue with the extension. Two of my oldest rules disappeared the last time I restarted the browser (just because). 

Maybe it doesn't have anything to do with the casing, but rather with the way the extension saves the settings.. maybe something relevant changed in one of the updates in the time frame between the moment you created that rule and today (that you edited it, or created a new one).

**Comment by Thorin-Oakenpants** [comment]
should really start a new issue for this, with STR, test pages to use, findings etc, and THEN unleash on that issue I created at Header Editor's reop

I will test something: the rule that ***was*** lowercase, I edited and saved as `Etag` and then it worked. I can disable this, because I created a new rule for lowercase. i.e - I can retest lowercase
^^ and nope, didn't work - so a brand new rule created and it didn't work. IMO IDK WTF TBH anymore.

I can disable all rules and create a new `ETag` one and see what happens - yup, that worked. Definitely case is the case (see what I did there)

**Comment by vertigo220** [comment]
I created three rules: etag, ETag, and eTaG. Only the first worked. It also worked with all of them enabled, so maybe a bunch of different rules (ETAG, etag, ETag, Etag, ...) could be made, or maybe it could be done with regex (something like [Ee|Tt|Aa|Gg]). Or would that possibly make it stand out and be more unique? Also, are you testing with the ghacks link from claustromaniac's post? I just assumed you were, but if not that could explain the difference.

**Comment by Thorin-Oakenpants** [comment]
> I created three rules: etag, ETag, and eTaG. Only the first worked

Is that a typo. Because the first one, all lowercase, is what has been on our wiki since we added it, and is the one that doesn't work for me. And I thought you said switching to `ETag` as per claustro's info then it did work.

**Comment by earthlng** [comment]
OMG! before you guys all go nuts I'll step in here :)

The problem looks like it's caused by a recent change in FF, not the extension. Not too long ago all headers were exposed by FF in exactly the casing sent by the server. Then they changed it to all lowercase for a while and now they apparently changed it back to original casing (IDK when exactly).
@vertigo220 what version of FF are you using?

But the extension code has also changed since the last time I looked at it, but frankly is still less than ideal. For example certain headers can exist multiple times and the extension only mods the 1st one.
For ETag it probably shouldn't matter but you never know. fe. IDK what FF does when a server sends 2 ETag headers.

The best solution that doesn't rely on FF keeping its current way of providing the headers nor any potential future changes in the extension itself, is to use a custom function. That should also work with all FF versions (ESR included) and all current, former and future releases of the extension.
And it bypasses the less than ideal implementation of the extension.

So, change your rule from Normal to Custom Function and paste this:
```js
for (let a in val) {
    if (val[a].name.toLowerCase() === 'etag') { val[a].value = ''; }
}
```
Don't use this in Chrome though because Chrome preserves empty headers and it might end up sending an empty If-None-Match header, IDK.

You'll know that it works when "Number of visits" never goes higher than `2` @ https://lucb1e.com/rp/cookielesscookies/

**Comment by earthlng** [comment]
Oh, and I forgot to mention that you can't rely on the devtools for headers because it doesn't accurately show changes made by extensions. In this case you'll see that it lists ETag in the response header despite it having been removed but you can deduce that the removal worked when you refresh and the next request header doesn't include If-None-Match. So despite being less than ideal the devtools are still somewhat useful because of the connection between these 2 headers.

**Comment by vertigo220** [comment]
Thanks @earthlng! That function is the only way, besides Secret Agent, that I've been able to pass the test at that site. And it's good to know it's more robust. Am I correct in understanding that code doesn't change the case, so etag=xyz will be set to etag="", ETag=xyz will be set to ETag="", and so on?

Also, I'm using Waterfox 56.2.0x64.

An OT question: could you or someone else tell me how I would make the following redirect work:

^http(s?)://www\.amazon\.com/gp/buy/spc/(.*)  >  https://smile.amazon.com/gp/buy/spc/handlers/display.html?

I figured out how to do it with Request Control, but since it seems Header Editor is capable of much more, I'm trying to transfer the rules from RC into it so I can reduce my number of addons. I've tried different variations, and the one I pasted here worked once (I think), but hasn't worked since.

**Comment by earthlng** [comment]
> Am I correct in understanding that code doesn't change the case, so etag=xyz will be set to etag="", ETag=xyz will be set to ETag="", and so on?

yes it doesn't change the case.

I never used Header Editor for redirects but I assume this should work:
match type: Regular Expression
match rules: `^https?://www.amazon.com/gp/buy/spc/.*`
redirect to: `https://smile.amazon.com/gp/buy/spc/handlers/display.html`

**Comment by vertigo220** [comment]
Thanks. I had my doubts, since it doesn't follow the "rules" in the examples from the wiki, but it does work, sort of. Unfortunately, it doesn't work when landing on the page, only when refreshing it, so I guess I'll have to ask on the HE repo.

**Comment by claustromaniac** [comment]
> OMG! before you guys all go nuts I'll step in here :)

If that includes me, I'm sorry to say you're too late. Like...years late.

Thanks for the explanation and the solution, mate! BTW did someone translate the HE documentation to english without me finding out? Last time I checked it was only in japanese IIRC. Do you know japanese? :stuck_out_tongue_closed_eyes: 

> I forgot to mention that you can't rely on the devtools for headers because it doesn't accurately show changes made by extensions. In this case you'll see that it lists ETag in the response header despite it having been removed but you can deduce that the removal worked when you refresh and the next request header doesn't include If-None-Match.

I pointed this out in my post above and everyone seemed to understand it. If my guess is right the inaccuracies should be in the response headers, especifically. Did you ever happen to dig into that too, by any chance? Just curious.

**Comment by claustromaniac** [comment]
BTW, one of my HE rules that disappeared from my list today just reappeared a while back. LOL! The other one is still MIA.

**Comment by Thorin-Oakenpants** [comment]
I wasn't exactly going nuts. Was going to test it with different FF and HE releases later on. What threw me a little was vertigo not stating that he's on Waterfox of all things - i.e essentially FF56 with god knows what changes under the hood by Mr. Alex. Unless someone states it, I will assume they are on current FF stable. I already suspected something was different re version with the diff in case with my FF60.

Thanks for clearing up that it was FF, and thanks for the solution re case.

Clastro - have you looked inside the HE storage?

**Comment by Thorin-Oakenpants** [comment]
Wiki updated

**Comment by claustromaniac** [comment]
> Clastro - have you looked inside the HE storage?

And what should I be looking for in there, exactly? I'll probably wipe it and rewrite the few rules that I use, anyway. It's not much of a hassle.

I'm still wondering why we got different results with the casing in our tests earlier. I only had issues with the lowercase `etag` rule when I tried the PoC, but both `etag` and `ETag` seemed to work for me everywhere else.. Then again, I didn't test much. I didn't stay around here for too long.

**Comment by Thorin-Oakenpants** [comment]
I have to admit that intrigued me as well (diff results), but now I no longer care. I've poked around the HE repo issues a few times, checking things out. The ones in English that is. I have no idea why you "lost" rules, I think I remember seeing some posts about it. It's never happened to me, I don't have any groups, and just one rule. But if it was a concern to you, because one "came back", then I suggested looking at the storage (you can peek inside these things - not sure what its using, localStorage or IDB of whatever) to see if it still existed. But if you're happy to just wipe the extension and UUID in total, and start over - be my guest. I'm not going to waste any more time on it.

**Comment by claustromaniac** [comment]
I'm not using groups either. I just had 5 rules and nothing really complicated. I wouldn't mind digging into the issue in other circumstances - and I probably will, if it turns out to be a recurring issue - but I'm pretty sure I'll invest less time rewriting them for the time being. 

Plus, I get to save _you_ the hassle of helping me with that! How nice is that? You sound tired enough as it is, :jeans:

**Comment by earthlng** [comment]
@claustromaniac 

> I pointed this out in my post above and everyone seemed to understand it.

yes I saw that, just wanted to re-confirm. And I'm not so sure *everyone* seemed to understand it :)

> If my guess is right the inaccuracies should be in the response headers, especifically. Did you ever happen to investigate that too, by any chance? Just curious.

No I did not. As far as I can tell the inaccuracies are all over the place

**Comment by claustromaniac** [comment]
> I'm not so sure everyone seemed to understand it :)

That's ok. I admit I often come short when I try to explain stuff to others anyway.

> No I did not. As far as I can tell the inaccuracies are all over the place

I've never seen inaccuracies in the request headers but I'll try to keep that in mind in case I ever do. I better stop trusting the devtools on that. Thanks again.

**Comment by vertigo220** [comment]
BTW, [here's another](https://robertheaton.com/2014/01/20/cookieless-user-tracking-for-douchebags/), unrelated cache tracking technique you may or may not be aware of which, AFAICT, can't be defeated except by periodically clearing the cache. Thought you might find it interesting.

**Comment by Thorin-Oakenpants** [comment]
> can't be defeated except by periodically clearing the cache

Or by not having any disk or memory cache. Or by Temporary Containers in a hardened mode. Or by not running JS (for that example, PoC)/  You could do this as well with a unique image name

**Comment by claustromaniac** [comment]
Thanks for sharing, @vertigo220. Another possibility is to prevent the browser from caching scripts, especifically, but I guess that approach is bound to break something sooner or later. There is obviously also a performance cost. 

In case you're curious, we could for example create another rule in Header Editor for modifying response headers with a custom function (like the one for ETags that @earthlng was so kind to show us), using something like this as the code:

```
if (detail.type === "script") {
    for (let a in val) {
        if (val[a].name.toLowerCase() === 'cache-control') { 
            val[a].value = 'no-cache, no-store, must-revalidate'; 
        }
    }
}
```

Can't recommend it for real daily use, though. I'm no guru. There might bad consequences for using something like this that I'm not aware of.

**Comment by Thorin-Oakenpants** [comment]
^ claustro - go repost that in `#436` which I just created. This thread needs retirement

**Comment by claustromaniac** [comment]
Done.

**Comment by Thorin-Oakenpants** [comment]
locking - if anyone has any etag question or issues, start a new issue
